KAFKA-17184: Fix the error thrown while accessing the RemoteIndexCache

kamalcph opened the pull request:
For segments that are uploaded to remote, RemoteIndexCache caches the
fetched offset, timestamp, and transaction index entries on the first
invocation to remote, then the subsequent invocations are accessed from
local.

The remote indexes that are cached locally gets removed on two cases:

1. Remote segments that are deleted due to breach by retention size/time
and start-offset.
2. The number of cached indexes exceed the remote-log-index-cache size
limit of 1 GB (default).

There are two layers of locks used in the RemoteIndexCache. First-layer
lock on the RemoteIndexCache and the second-layer lock on the
RemoteIndexCache#Entry.

**Issue**

1. The first-layer of lock coordinates the remote-log reader and deleter
threads. To ensure that the reader and deleter threads are not blocked
on each other, we only take `lock.readLock()` when accessing/deleting
the cached index entries.
2. The issue happens when both the reader and deleter threads took the
readLock, then the deleter thread marked the index as
`markedForCleanup`. Now, the reader thread which holds the `indexEntry`
gets an IllegalStateException when accessing it.
3. This is a concurrency issue, where we mark the entry as
`markedForCleanup` before removing it from the cache. See
RemoteIndexCache#remove, and RemoteIndexCache#removeAll methods.
4. When an entry gets evicted from cache due to breach by maxSize of 1
GB, then the cache remove that entry before calling the evictionListener
and all the operations are performed atomically by caffeine cache.

**Solution**

1. When the deleter thread marks an Entry for deletion, then we rename
the underlying index files with ".deleted" as suffix and add a job to
the remote-log-index-cleaner thread which perform the actual cleanup.
Previously, the indexes were not accessible once it was marked for
deletion. Now, we allow to access those renamed files (from entry that
is about to be removed and held by reader thread) until those relevant
files are removed from disk.
2. Similar to local-log index/segment deletion, once the files gets
renamed with ".deleted" as suffix then the actual deletion of file
happens after `file.delete.delay.ms` delay of 1 minute. The renamed
index files gets deleted after 10 seconds.
3. During this time, if the same index entry gets fetched again from
remote, then it does not have conflict with the deleted entry as the
file names are different.

Reviewers: Satish Duggana <satishd@apache.org>

kamalcph committed:
KAFKA-17184: Fix the error thrown while looking up offset in RemoteIndexCache

For segments that are uploaded to remote, RemoteIndexCache caches the fetched offset, timestamp, and transaction index entries on the first invocation to remote, then the subsequent invocations are accessed from local.

The remote indexes that are cached locally gets removed on two cases:

1. Remote segments are deleted due to breach by retention size/time and start-offset.
2. The number of cached indexes exceed the remote-log-index-cache size limit of 1 GB (default).

There are two layers of locks used in the RemoteIndexCache. One first-layer lock on the RemoteIndexCache and the second-layer lock on the RemoteIndexCache#Entry.

Issue:

1. The first-layer of lock coordinates the remote-log reader and deleter threads. To ensure that the reader and deleter threads are not blocked on each other, we only take `lock.readLock()` when accessing/deleting a cached index entries.
2. The issue happens when both the reader and deleter threads took the readLock, then the deleter thread marked the index as `markedForCleanup`. Now, the reader thread which holds the `indexEntry` gets an IllegalStateException when accessing it.
3. This is a concurrency issue, where we mark the entry as `markedForCleanup` before removing it from the cache. See RemoteIndexCache#remove, and RemoteIndexCache#removeAll methods.
4. When an entry gets evicted from cache, then the cache remove that entry before calling the evictionListener and all the operations are performed atomically by caffeine cache.

Solution:

1. When the deleter thread marks an Entry for deletion, then we rename the underlying index files with ".deleted" as suffix and add a job to the remote-log-index-cleaner thread which perform the actual cleanup. Previously, the indexes were not accessible once it was marked for deletion. Now, we are allowing to access thoes renamed files until they are removed from disk.
2. Similar to local-log index/segemnt deletion, once the files gets renamed with ".deleted" as suffix then the actual deletion of file happens after file.delete.delay.ms delay of 1 minute.
3. During this time, if the same index entry gets fetched again from remote, then it does not have conflict with the deleted entry as the file names are different.

- Refactor RemoteIndexCache and its tests
- Reduced the fileDeleteDelayMs from 60s to 10s
- updated the unit tests

kamalcph commented:
Ping for review. I've made two commits in the PR to make the review easier:

1.  81f81d9f397906514af136d8d3b37ffc052b6d66 - refactoring
2. 00479eb90fa8de04e8073b4b04d00d9a692bb55b - actual fix

PTAL. 

cc @jeqo @showuon @satishd @dopuskh3

satishd commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -91,18 +90,14 @@ public class RemoteIndexCache implements Closeable {
     private final RemoteStorageManager remoteStorageManager;
-    private final ShutdownableThread cleanerThread;
+    private final KafkaScheduler cleanerScheduler = new KafkaScheduler(1, true, REMOTE_LOG_INDEX_CACHE_CLEANER_THREAD);
+    private int fileDeleteDelayMs = 60_000;
```
I think the value of 60 secs looks to be reasonable. We can revisit setting a config if really needed in future.

satishd commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -208,46 +203,32 @@ public void removeAll(Collection<Uuid> keys) {
+                    log.error("Error occurred while cleaning up expired entry", ex);
+                }
+            };
+            cleanerScheduler.scheduleOnce("delete-index", runnable, fileDeleteDelayMs);
```
Good catch on the earlier issue and fixing it using a schedule delay like we do with local segments.

satishd reviewed (commented):
Thanks @kamalcph for the PR. Nice cleanup and good catch! Overall LGTM.

Can you rebase and resolve the conflicts in the test class?

kamalcph commented:
Rebased the PR with trunk. In #19474, while rewriting the tests from scala to Java, some recent updates were not taken. Added those left-over unit test with this PR: 

1. testDeleteInvalidIndexFilesOnInit() -> existing unit test missed while porting to java
2. testIndexFileAlreadyExistOnDiskButNotInCache() -> test was updated previously, but the changes were not taken while converting to java. Also, fixed this unit test.

kamalcph committed:
update unit tests

kamalcph committed:
Add unit test for scheduler#pendingTaskSize

kamalcph committed:
Change the method modifier to package-protected as the test class is moved to same package

satishd reviewed (approved):
Thanks @kamalcph resolving the conflicts, added the missing tests in the earlier PR, and addressing the comments that we had through. LGTM.

chia7712 commented:
@kamalcph the new test `testPendingTaskSize` could be hanging (see https://github.com/apache/kafka/actions/runs/14560322254/job/40842877645?pr=19521), and I have opened https://issues.apache.org/jira/browse/KAFKA-19180 to trace it. I will file a PR later
